lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,1039 @@
+		default:
+		}
+
+		delete(s.responses, requestID)
+	}
+}
+
+// getResponse waits for a response with the given request ID and returns the
+// response if it is received. If the corresponding response from the remote
+// signer is a SignerError, the error message is returned. If the response is
+// not received within the given timeout, an error is returned.
+//
+// Note: Before calling this function, the caller must have created a response
+// channel for the request ID.
+func (s *SignCoordinator) getResponse(requestID uint64,
+	timeout time.Duration) (*walletrpc.SignCoordinatorResponse, error) {
+
+	s.mu.Lock()
+
+	// Verify that we have a response channel for the request ID.
+	if _, ok := s.responses[requestID]; !ok {
+		// It should be impossible to reach this case, as we create the
+		// response channel before sending the request.
+		s.mu.Unlock()
+
+		return nil, fmt.Errorf("no response channel found for "+
+			"request ID %d", requestID)
+	}
+
+	// Grab the response channel for the request ID.
+	respChan := s.responses[requestID]
+
+	s.mu.Unlock()
+
+	// Wait for the response to arrive.
+	select {
+	case resp, ok := <-respChan:
+		if !ok {
+			// If the response channel was closed, we return an
+			// error as the receiving thread must have timed out
+			// before we managed to grab the response.
+			return nil, ErrRequestTimeout
+		}
+
+		// If the response is an error, we return the error message.
+		//nolint:lll
+		if errorResp, ok := resp.GetSignResponseType().(*walletrpc.SignCoordinatorResponse_SignerError); ok {
+			errStr := errorResp.SignerError.GetError()
+
+			log.Debugf("Received an error response from remote "+
+				"signer for request ID %d. Error: %v",
+				requestID, errStr)
+
+			return nil, errors.New(errStr)
+		}
+
+		log.Debugf("Received remote signer response for request ID %d",
+			requestID)
+
+		return resp, nil
+
+	case <-s.doneReceiving:
+		log.Debugf("Stopped waiting for remote signer response for "+
+			"request ID %d as the stream has been closed",
+			requestID)
+
+		return nil, ErrNotConnected
+
+	case <-s.quit:
+		log.Debugf("Stopped waiting for remote signer response for "+
+			"request ID %d as we're shutting down", requestID)
+
+		return nil, ErrShuttingDown
+
+	case <-time.After(timeout):
+		log.Debugf("Remote signer response timed out for request ID %d",
+			requestID)
+
+		return nil, ErrRequestTimeout
+	}
+}
+
+// registerRequest registers a new request with the SignCoordinator, ensuring it
+// awaits the handling of the request before shutting down. The function returns
+// a Done function that must be executed once the request has been handled.
+func (s *SignCoordinator) registerRequest() (func(), error) {
+	// We lock the mutex to ensure that we can't have a race where we'd
+	// register a request while shutting down.
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	select {
+	case <-s.quit:
+		return nil, ErrShuttingDown
+	default:
+	}
+
+	s.wg.Add(1)
+
+	return func() {
+		s.wg.Done()
+	}, nil
```

Comment:
Hmm, there are a lot of layers with the mutexes and wait groups... Probably need to take another, closer look at those, to make sure we can't run into deadlocks...